javascript_nodejs/50.contoso-cafe-bot/bot.js

```diff
@@ -0,0 +1,151 @@
+                case ActivityTypes.Message: {
+                    // Process on turn input (card or NLP) and gather new properties
+                    // OnTurnProperty object has processed information from the input message activity.
+                    let onTurnProperties = await this.detectIntentAndEntities(context);
+                    if(onTurnProperties === undefined) break;
+                    
+                    // Set the state with gathered properties (intent/ entities) through the onTurnPropertyAccessor
+                    this.onTurnPropertyAccessor.set(context, onTurnProperties);
+                    
+                    // Do we have any oustanding dialogs? if so, continue them and get results
+                    // No active dialog? start a new main dialog
+                    await this.continueOrBeginMainDialog(context);
+                    break;
+                }
+                case ActivityTypes.ConversationUpdate: {
+                    if (context.activity.membersAdded.length !== 0) {
+                        // Iterate over all new members added to the conversation
+                        // Greet anyone that was not the target (recipient) of this message
+                        // the 'bot' is the recipient for events from the channel,
+                        // turnContext.Activity.MembersAdded == turnContext.Activity.Recipient.Id indicates the
+                        // bot was added to the conversation.
+                        // TODO: Send welcome card the right way
+                        if(context.activity.membersAdded[0].name !== 'Bot') await this.welcomeUser(context);
+                    }
+                    break;
+                }
+                default: {
+                    // Handle other activity types as needed.
+                    break;            
+                }
+            }
+        }
+        /**
+         * Async method to continue or begin main dialog
+         * 
+         * @param {Object} context conversation context object
+         * 
+         */
+        async continueOrBeginMainDialog(context) {
+            // Create dialog context.
+            const dc = await this.dialogs.createContext(context);
+            
+            // Continue outstanding dialogs. 
+            const result = await dc.continue();
+            
+            // If no oustanding dialogs, begin main dialog
+            if (result.status === DialogTurnStatus.empty) {
+                await dc.begin(MainDialog.name);
+            }
+        }
+        /**
+         * Async helper method to get on turn properties from cards or NLU using https://LUIS.ai
+         * 
+         * @param {Object} context conversation context object
+         * 
+         */
+        async detectIntentAndEntities (context) {
+            // Handle card input (if any), update state and return.
+            if (context.activity.value !== undefined) return OnTurnProperty.fromCardInput(context.activity.value);
```

Comment:
>OnTurnProperty [](start = 61, length = 14)

What is OnTurnProperty?  This name is super confusing.  Is this an event callback?  A class name?